### PR TITLE
[meshcop] implement a prioritized queue for Joiner Routers

### PR DESCRIPTION
**Pull request by CiaranWoodward**
In relation to #2136 

Implements a priority queue which the joiner will iterate through when attempting to join a network.

### CODE REVIEW

File: src/core/meshcop/joiner.cpp [hunk 1]
```diff
@@ -201,57 +203,102 @@ void Joiner::HandleDiscoverResult(otActiveScanResult *aResult)
...
+
+exit:
+    otLogFuncExit();
+}
+
+void Joiner::AddJoinerRouter(JoinerRouter &aJoinerRouter)
+{
+    JoinerRouter *joinerRouter = &mJoinerRouters[0];
+
+    //Find the lowest priority
```
Comment (jwhui) [inline src/core/meshcop/joiner.cpp]: nit: space after `//`

File: src/core/meshcop/joiner.cpp [hunk 2]
```diff
@@ -201,57 +203,102 @@ void Joiner::HandleDiscoverResult(otActiveScanResult *aResult)
...
         {
-            VerifyOrExit(!mJoinerRouterIsSpecific);
-            VerifyOrExit(aResult->mRssi > mJoinerRouterRssi);
+            otLogDebgMeshCoP(GetInstance(), "Steering data does not include this device");
+            ExitNow();
         }
-        else if (steeringData.GetBit(mCcitt % steeringData.GetNumBits()) &&
-                 steeringData.GetBit(mAnsi % steeringData.GetNumBits()))
+
+        joinerRouter.mPriority = (uint16_t)((int16_t) aResult->mRssi + 0x80);
```
Comment (jwhui) [inline src/core/meshcop/joiner.cpp]: nit: prefer using `static_cast()`.

File: src/core/meshcop/joiner.cpp [hunk 3]
```diff
@@ -181,8 +184,7 @@ void Joiner::HandleDiscoverResult(otActiveScanResult *aResult, void *aContext)
 
 void Joiner::HandleDiscoverResult(otActiveScanResult *aResult)
 {
-    ThreadNetif &netif = GetNetif();
-    Ip6::MessageInfo messageInfo;
+    JoinerRouter joinerRouter;
```
Comment (bukepo) [inline src/core/meshcop/joiner.cpp]: Move this definition into the if block to reduce the scope?

File: src/core/meshcop/joiner.cpp [hunk 4]
```diff
@@ -201,57 +203,102 @@ void Joiner::HandleDiscoverResult(otActiveScanResult *aResult)
...
+}
 
-        netif.GetMac().SetPanId(mJoinerRouterPanId);
-        netif.GetMac().SetChannel(mJoinerRouterChannel);
+otError Joiner::TryNextJoin()
+{
+    ThreadNetif &netif = GetNetif();
+    otError error = OT_ERROR_NOT_FOUND;
+    JoinerRouter *joinerRouter = &mJoinerRouters[0];
+    Ip6::MessageInfo messageInfo;
```
Comment (bukepo) [inline src/core/meshcop/joiner.cpp]: move this into the if block?

File: src/core/meshcop/joiner.hpp [hunk 5]
```diff
@@ -96,6 +96,16 @@ class Joiner: public ThreadNetifLocator
...
+        kSpecificPriorityBonus = (1 << 9),
+    };
+
+    struct JoinerRouter
+    {
+        uint16_t mPriority;
+        uint16_t mPanId;
+        uint16_t mJoinerUdpPort;
+        uint8_t mChannel;
+        Mac::ExtAddress mExtAddr;
```
Comment (bukepo) [inline src/core/meshcop/joiner.hpp]: make the 8 byte `mExtAddr` the first member?

**Comment by CiaranWoodward** [comment]
Thanks for all of the reviews, I fixed the points mentioned and rebased on top of master to fix merge errors

File: src/core/openthread-core-default-config.h [hunk 6]
```diff
@@ -306,6 +306,16 @@
 #endif  // OPENTHREAD_CONFIG_MAX_JOINER_ENTRIES
 
 /**
+ * @def OPENTHREAD_CONFIG_MAX_JOINER_ENTRIES
+ *
+ * The maximum number of Joiner Router entries that can be queued by the Joiner.
+ *
+ */
+#ifndef OPENTHREAD_CONFIG_MAX_JOINER_ROUTER_ENTRIES
+#define OPENTHREAD_CONFIG_MAX_JOINER_ROUTER_ENTRIES                    2
```
Comment (bukepo) [inline src/core/openthread-core-default-config.h]: this `2` not aligned with other definition.

Comment (CiaranWoodward) [reply src/core/openthread-core-default-config.h]: fixed

**Pull request merged**